stdlib/SparseArrays/docs/src/AbstractSparseMatrixCSC.md

```diff
@@ -0,0 +1,59 @@
+    AbstractSparseMatrixCSC{Tv,Ti<:Integer} <: AbstractSparseMatrix{Tv,Ti}
+
+Supertype for matrix with compressed sparse column (CSC).
+
+## `AbstractSparseMatrixCSC` interface
+
+In addition to the [Abstract array interface](@ref man-interface-array), every
+`AbstractSparseMatrixCSC` subtype `TS` must provide the following methods:
+
+* [`size(::TS)`](@ref size)
+* [`getcolptr(::TS) :: AbstractVector{<:Ti}`](@ref getcolptr)
+* [`rowvals(::TS) :: AbstractVector{<:Ti}`](@ref rowvals)
+* [`nonzeros(::TS) :: AbstractVector{<:Tv}`](@ref nonzeros)
+
+Other sparse matrix methods such as [`nzrange`](@ref) and [`nnz`](@ref) are automatically
+defined in terms of above functions.
+
+## Assumed invariance
+
+To use algorithms defined in SparseArrays, a matrix `A` of type `AbstractSparseMatrixCSC`
+must satisfy the following constraints.
+
+### Matrix `A` and all vectors constituting `A` have one-based indexing
+
+```julia
+@assert !has_offset_axes(A)
+@assert !has_offset_axes(getcolptr(A))
+@assert !has_offset_axes(rowvals(A))
+@assert !has_offset_axes(nonzeros(A))
+```
+
+### Row indices in `rowval(A)` for each column are sorted
+
+```julia
+for col in axes(A, 2)
+    @assert issorted(rowval(A)[nzrange(A, col)])
+end
+```
+
+### Column pointers in `getcolptr(A)` are increasing, started at 1
+
+```julia
+@assert getcolptr(A)[1] === 1
+@assert all(diff(getcolptr(A)) .>= 0)
+```
+
+### Row index vector `rowvals(A)` and non-zero value vector `nonzeros(A)` are long enough
+
+```julia
+@assert nnz(A) <= length(rowvals(A))
+@assert nnz(A) <= length(nonzeros(A))
+```
+
+### Indices for rows, `rowvals(A)` and `nonzeros(A)` are representable by `Ti`
+
+```julia
+@assert size(A, 1) <= typemax(Ti)
+@assert nnz(A) <= typemax(Ti)
```

Comment:
So this is something I don't understand in #31724.  It was mentioned that both `m = size(A, 1)` and `n = size(A, 2)` must be representable by `Ti`:

https://github.com/JuliaLang/julia/blob/fb9cd344a40df778aeb8f9eb73aa4f19a16d8406/HISTORY.md#L88-L90

I understand that `size(A, 1) <= typemax(Ti)` has to hold because we use `rowvals(A)` to indexing to the row indices [1].  However, `size(A, 2)` only appears as the indexing into `getcolptr(S)` so why does `Ti` matter?

Also regarding the current comment:

https://github.com/JuliaLang/julia/blob/19a2b2e417fe80e83e2d2964db70451835abb4d4/stdlib/SparseArrays/src/sparsematrix.jl#L8-L9

* It says `nnz(A) < typemax(Ti)`. But isn't `nnz(A) == typemax(Ti)` OK?
* I suppose `length(rowvals(A)) <= typemax(Ti)` and `length(nonzeros(A)) <= typemax(Ti)` don't have to hold because the tail of those vectors can just be ignored?

---

[1] But technically `size(A, 1) > typemax(Ti)` sounds OK as long as the structural nonzeros appear only in the upper `1:typemax(Ti)` rows?